java/src/main/java/com/genexus/db/driver/GXPreparedStatement.java

```diff
@@ -934,89 +933,99 @@ public void setGXDbFileURI(int index, String fileName, String blobPath, int leng
+			setVarchar(index, fileName, length, false);
+			return;
+		}
+
+		String fileUri = "";
+    	String multimediaTemporalPath = com.genexus.Preferences.getDefaultPreferences().getMultimediaPath();
+
+		if (fileName.length() > 0 && !(blobPath.startsWith(multimediaTemporalPath) && GXutil.isAbsoluteURL(fileName))) {
+			fileUri = GXDbFile.generateUri(fileName, !GXDbFile.hasToken(fileName), true);
+		}
+		else if (blobPath.trim().length() > 0)
+		{
+			blobPath = com.genexus.GXutil.cutUploadPrefix(blobPath);
+			File file = new File(blobPath);
+			fileUri = GXDbFile.generateUri(file.getName(), !GXDbFile.hasToken(blobPath), true);
+		}
+
+		ExternalProvider storageProvider = Application.getExternalProvider();
+		boolean externalStorageEnabled = storageProvider != null;
+		boolean attInExternalStorage = externalStorageEnabled && (tableName != null && fieldName != null); //Should improve this condition in order to not depend on tableName and FieldName.
+
+		if (!attInExternalStorage) {
+			setVarchar(index, fileUri, length, false);
+			return;
+		}
+
+		//External Storage is ENABLED. We have the following cases:
+		// - 1. WebUpload: The URL is an External Storage URL in the Private Temp Storage Folder.
+		// - 2. External URL: An absolute URL (outside External Storage).
+		// - 3. Transaction Update Mode (nothing changed): An Storage URL that is already in External Storage (except in private), should not be uploaded nor copied.
+		// - 4. Upload Resource from local drive to External Storage: Ex: Image.FromImage(ActionDelete)
+
+		String storageObjectName = ExternalProviderCommon.getProviderObjectName(storageProvider, blobPath);
+		boolean resourceAlreadyOnStorage = storageObjectName != null;
+		ResourceAccessControlList defaultAcl = ResourceAccessControlList.Default;
+
+		String storageTargetObjectName = blobPath;
+		boolean isPrivateTempUpload = storageTargetObjectName.startsWith(Application.getClientPreferences().getTMPMEDIA_DIR());
+		String folder = Application.getGXServices().get(GXServices.STORAGE_SERVICE).getProperties().get("FOLDER_NAME");
+		if (!resourceAlreadyOnStorage && (fileName.equals(blobPath) || GXutil.isAbsoluteURL(fileName))) {
+			// - 2. External URL: An absolute URL (outside External Storage).
+			try (InputStream is = new URL(fileName).openStream()){
+				int idx = storageTargetObjectName.lastIndexOf("/");
+				if ((idx != -1) && (idx < storageTargetObjectName.length() - 1)) {
+					storageTargetObjectName = storageTargetObjectName.substring(idx + 1);
 				}
+				storageTargetObjectName = com.genexus.PrivateUtilities.getTempFileName("", CommonUtil.getFileName(storageTargetObjectName), CommonUtil.getFileType(storageTargetObjectName), true);
+				storageTargetObjectName = folder + "/" + tableName + "/" + fieldName + "/" + storageTargetObjectName;
+				fileUri = Application.getExternalProvider().upload(storageTargetObjectName, is, defaultAcl);
+			} catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)